### PR TITLE
Iox #340 simplify expected optionals

### PR DESCRIPTION
**Pull request by ithier**
* adds an `and_then` specialization for `cxx::expected<cxx::optional<T>>` that automatically retrieves the value from non-empty optionals
* the and_then specialization does nothing if the optional is empty
* adds `if_empty()` method to `cxx::expected<cxx::optional<T>>` types that allows users to react to retrieving an empty optional without error

### CODE REVIEW

**Comment by elBoberido** [comment]
@elfenpiff told me yesterday about an idea which makes your proposal more generic, maybe you can discuss with him about that. I'm fine with either solution

**Comment by ithier** [comment]
> @elfenpiff told me yesterday about an idea which makes your proposal more generic, maybe you can discuss with him about that. I'm fine with either solution

@elBoberido I spoke briefly with him about it. I would still like to merge this as a stepping stone towards the generic solution.

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl [hunk 1]
```diff
@@ -346,6 +346,58 @@ expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(ValueType&
     return *this;
 }
 
+template <typename ValueType, typename ErrorType>
+template <typename Optional = ValueType, typename std::enable_if<is_optional<Optional>::value, int>::type = 0>
```
Comment (mossmaurice) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:350]: @elfenpiff Can the 2nd template parameter be omitted as it's already in the hpp? I remember we did in some first version of the `WaitSet`.

**Review by mossmaurice — APPROVED** [review]
LGTM

**Review by mossmaurice — CHANGES_REQUESTED** [review]
Fix build error related to 2nd template param in new `expected` methods

**Comment by elBoberido** [comment]
> Fix build error related to 2nd template param in new `expected` methods

and there is also this warning in the Colcon build check
```
The `set-env` command is deprecated and will be disabled on November 16th. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
```
Not related to this PR, but we should fix it the next few days.

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_cxx_expected.cpp [hunk 2]
```diff
@@ -248,3 +257,51 @@ TEST_F(expected_test, ExpectedWithErrorConvertsToNullopt)
     optional<int> value = sut.to_optional();
     ASSERT_THAT(value.has_value(), Eq(false));
 }
+
+TEST_F(expected_test, AndThenUnpacksOptionalWhenNonEmptyOptionalValue)
+{
+    auto sut = expected<iox::cxx::optional<int>, float>::create_value(123);
+    MockCallables mocks{};
+    EXPECT_CALL(mocks, onSuccess).Times(1);
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_expected.cpp:265]: I think you could get rid of `MockCallables` completely, reduce the code and test everything anyway.
Look here:
```cpp
bool andThenWasCalled = false;
sut.and_then([&andThenWasCalled](int& val) {
        andThenWasCalled = true;
        ASSERT_THAT(val, Eq(123));
    });
EXPECT_TRUE(andThenWasCalled);
```

You could also create an integer which is incremented every time you call the the lambda to verify that it is only called once.

**Review by elfenpiff — APPROVED** [review]
@ithier If you have time you could maybe look again over the tests but it is mandatory...

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl [hunk 3]
```diff
@@ -346,6 +346,58 @@ expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(ValueType&
...
+{
+    return const_cast<expected*>(this)->and_then(callable);
+}
+
+template <typename ValueType, typename ErrorType>
+template <typename Optional, typename std::enable_if<is_optional<Optional>::value, int>::type>
+inline expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(typename Optional::type&)>& callable) noexcept
+{
+    if (!this->has_error())
```
Comment (MatthiasKillat) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:363]: Can we add a has_value function to the expected in this pull request? I know that it is not strictly needed, but I would prefer to have both (has_error and has_value) or only has_value for consistency.

If we do so, we could use it here.

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl [hunk 4]
```diff
@@ -346,6 +346,58 @@ expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(ValueType&
...
+template <typename Optional, typename std::enable_if<is_optional<Optional>::value, int>::type>
+inline const expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(typename Optional::type&)>& callable) const
+    noexcept
+{
+    return const_cast<expected*>(this)->and_then(callable);
+}
+
+template <typename ValueType, typename ErrorType>
+template <typename Optional, typename std::enable_if<is_optional<Optional>::value, int>::type>
```
Comment (MatthiasKillat) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:359]: There is also enable_if_t (C++14) which can be used to shorten this a bit (not required).

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl (hunk 4 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:359]: This was already added to our cxx/type_trait.hpp and can be removed once we switched to C++14

File: iceoryx_utils/test/moduletests/test_cxx_expected.cpp (hunk 2 above)
Comment (ithier) [reply iceoryx_utils/test/moduletests/test_cxx_expected.cpp:265]: Will get it in the next PR.

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl (hunk 3 above)
Comment (ithier) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:363]: Will get it in a follow-up.

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl (hunk 4 above)
Comment (ithier) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl:359]: Ah, I forgot about this.

**Pull request merged**